### PR TITLE
Store MetaData for DiskDatasets as hd5 File

### PR DESCRIPTION
**Pull request by lilleswing**
The non-compatability of our DiskDataset was because we were storing pickled pandas dataframes as the metadata on disk.

This CR moves storing this data to a hd5 file instead which should be stable moving forward.  It lazily updates existing Datasets in place on read when they are using the old format.

Another object that was stored as a pickle was a list of strings representing the tasks.  Since this is small I store it in a json file.

### CODE REVIEW

File: deepchem/data/datasets.py [hunk 1]
```diff
@@ -464,12 +462,27 @@ def create_dataset(shard_generator, data_dir=None, tasks=[], verbose=True):
...
+    try:
+      tasks_filename, metadata_filename = self._get_metadata_filename()
+      with open(tasks_filename) as fin:
+        tasks = json.load(fin)
+      metadata_df = read_hdf(metadata_filename, 'metadata')
+      return tasks, metadata_df
+    except Exception as e:
+      pass
+
+    metadata_filename = os.path.join(self.data_dir, "metadata.joblib")
```
Comment (peastman) [inline deepchem/data/datasets.py]: Maybe add a comment at this point to explain what it's doing?  "The metadata is stored in an obsolete format, so upgrade it to the newer format."

Comment (rbharath) [reply deepchem/data/datasets.py]: +1

**Comment by rbharath** [comment]
@lilleswing Like the changes! The pickled dataframes have been a consistent source of headaches so happy to see the changes.

**Comment by rbharath** [comment]
@lilleswing Is this ready to merge in?

**Comment by lilleswing** [comment]
Don't merge in yet -- I want to do more manual testing about version compatibilities.

**Comment by lilleswing** [comment]
I ran my tests -- good to go when green now.

**Comment by rbharath** [comment]
@lilleswing Is this ready to merge in?

**Comment by lilleswing** [comment]
ready to go.  I want to make some cachine enhancements for shapes of Diskdataset on top of this.  But that can go in on either side of the cut.

**Comment by lilleswing** [comment]
Merge Away

**Comment by rbharath** [comment]
LGTM

**Pull request merged**